### DISCUSSION TITLE
discord picture send

### DISCUSSION THREAD

**Discussion by olikann** [discord-jda/JDA#2484] [Questions and Help]
Hello everyone! Can someone help me with how to use the .sendFile and .addFile methods in JDA 5.0.0-beta11? I want to send an image file to a channel, but when I use .addFile and .sendFile, it gives me an underline error saying "cannot find symbol". Has anything changed regarding these methods?

**Comment by MinnDevelopment** [comment]
The [migration guide](https://jda.wiki/introduction/migration-v4-v5/#file-sending) explains:

> The old `sendFile(...)`/`replyFile(...)` overloads available on MessageChannel and interactions has been replaced by a single `sendFiles(...)` method. This new method accepts the `FileUpload` type, which also supports file descriptions (alt text) via `setDescription(...)`. For example, old code such as `sendFile(data, name, AttachmentOption.SPOILER)` is replaced with `sendFiles(FileUpload.fromData(data, name).asSpoiler())`.
> 
> Due to changes in the Discord API v10, you can no longer add files to messages without replacing all existing attachments. We could previously support methods such as `MessageAction#addFile` for editing messages, which is no longer possible in API v10. Instead, you must replace the entire list of attachments, using `MessageEditAction#setAttachments`. Here is an example:
> 
> ```java
> // Here "message" is an instance of the Message interface
> 
> // Take the first attachment of the message, all others will be removed
> AttachedFile attachment = message.getAttachments().get(0);
> 
> // The name here will be "cat.png" to discord, what the file is called on your computer is irrelevant and only used to read the data of the image.
> FileUpload file = FileUpload.fromData(new File("mycat-final-copy.png"), "cat.png"); // Opens the file called "cat.png" and provides the data used for sending
> 
> // Edit request to keep the first attachment, and add one more file to the message
> message.editMessage("New content")
>        .setAttachments(attachment, file)
>        .queue();
> ```

**Comment by olikann** [comment]
Hi! Thank you for the information. However, could you help me with something? I've created a Discord bot using the latest version of JDA, and I've also made a server.js file for generating a welcome card image. Now, I'm trying to send that generated image to a specific Discord channel within the GuildMemberJoin method, but I'm facing difficulties. The generated image appears correctly in my browser, but I'm unsure how to send it to the Discord channel because methods like .addFile are not working. If it helps, I can share the code with you. Thank you in advance for your assistance. how can i send this picture to the channel because it work in my browser: 

example link:

[http://localhost:8080/generateCard?username=olikan&avatar=https://is3-ssl.mzstatic.com/image/thumb/Purple116/v4/34/95/4d/34954db6-2955-2514-e9e5-5edb3ec66bdb/AppIcon-0-0-1x_U007emarketing-0-0-0-7-0-0-sRGB-0-0-0-GLES2_U002c0-512MB-85-220-0-0.png/1200x630wa.png&memberCount=11](http://localhost:8080/generateCard?username=olikan&avatar=https%3A%2F%2Fis3-ssl.mzstatic.com%2Fimage%2Fthumb%2FPurple116%2Fv4%2F34%2F95%2F4d%2F34954db6-2955-2514-e9e5-5edb3ec66bdb%2FAppIcon-0-0-1x_U007emarketing-0-0-0-7-0-0-sRGB-0-0-0-GLES2_U002c0-512MB-85-220-0-0.png%2F1200x630wa.png&memberCount=11)

this us my server.js file: 

const express = require("express");
const cors = require("cors");
const { welcomeImage } = require("discord-welcome-card");

const app = express();
app.use(cors());

app.get("/generateCard", async (req, res) => {
    //if(!req.query["username"] || !req.query["avatar"] || !req.query["memeberCount"]){
     //   return res.status(400).send();
    //}
    const member = {
        user : {
            tag : req.query["username"],
            displayAvatarURL : () => req.query["avatar"],
            
        },
        guild : {
            memberCount : req.query["memberCount"]
        }
    };
    const image = await welcomeImage(member, {theme : "dark"});
    res.contentType("image/png");
    res.send(image);
});
app.listen(8080);








and this is my java code atm: 
public String sendGETRequest(String url) throws IOException {
        URL obj = new URL(url);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();
        con.setRequestMethod("GET");

        int responseCode = con.getResponseCode();
        if (responseCode == HttpURLConnection.HTTP_OK) {
            BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
            String inputLine;
            StringBuilder response = new StringBuilder();

            while ((inputLine = in.readLine()) != null) {
                response.append(inputLine);
            }
            in.close();

            return response.toString();
        } else {
            throw new IOException("GET request failed. Response Code: " + responseCode);
        }
    }

    @Override
    public void onGuildMemberJoin(GuildMemberJoinEvent event) {
        
        Guild guild = event.getGuild();
        try {
            String username = event.getMember().getUser().getName();
            String avatar = event.getMember().getUser().getAvatarUrl();
            if (avatar == null || avatar.isEmpty()) {
                avatar = "https://assets-global.website-files.com/6257adef93867e50d84d30e2/636e0a6a49cf127bf92de1e2_icon_clyde_blurple_RGB.png"; //ha nem találja a faszt
            }
            int memberCount = event.getGuild().getMemberCount();

            URIBuilder uriBuilder = new URIBuilder("http://localhost:8080/generateCard");
            uriBuilder.addParameter("username", username);
            uriBuilder.addParameter("avatar", avatar);
            uriBuilder.addParameter("memberCount", String.valueOf(memberCount));

            String url = uriBuilder.toString();

            String imageUrl = "http://localhost:8080/generateCard?username=" + username
                    + "&avatar=" + avatar
                    + "&memberCount=" + memberCount;
            
            
            URL imageUrlObject = new URL(imageUrl);
            BufferedImage image = ImageIO.read(imageUrlObject);
            
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            ImageIO.write(image, "png", baos);
            InputStream inputStream = new ByteArrayInputStream(baos.toByteArray());

            TextChannel channel = guild.getTextChannelById("927462788132139088");
            if (channel != null) {
// here need the code......................

            }

        } catch (IOException e) {
            e.printStackTrace();
        } catch (URISyntaxException ex) {
            Logger.getLogger(Listeners.class.getName()).log(Level.SEVERE, null, ex);
        }
}

thanks
Zoli

**Reply by MinnDevelopment** [reply]
I've already provided example code in my answer.

**Reply by olikann** [reply]
Hi again! Alright, I tried to write the code based on the given code snippet, but the problem is that the server.js generates the welcome card, which is displayed in the browser. I managed to store it in memory using the ImageIO.write method, but I cannot send it from there because FileUpload.fromData doesn't allow me to do that, if I understand correctly. Specifically, this line: FileUpload file = FileUpload.fromData(imageFile, "card.png"). The message.editMessage() method doesn't directly support attaching image files from memory. Can you suggest something?

**Reply by MinnDevelopment** [reply]
As explained in the first line from my quote, you can use `sendFiles(...)` with the `FileUpload` instance. `FileUpload.fromData(...)` can also be used with `InputStream` or `byte[]` instead of `File`.
